src/python/pants/engine/target.py

```diff
@@ -124,20 +101,37 @@ def compute_value(cls, raw_value: Optional[int], *, address: Address) -> Optiona
                 return value_or_default
     """
 
-    value: ImmutableValue
+    # This exists on every Field subclass.
+    value: _DefaultBase
+    # Subclasses must define these.
+    alias: ClassVar[str]
+    default: ClassVar[_DefaultBase]
+    # Subclasses may define these.
+    # TODO: enforce that `raw_value` and `value` are *not* Optional if `required = True`, possibly
+    # via mixin or decorator?
+    required: ClassVar[bool] = False
+    deprecated_removal_version: ClassVar[Optional[str]] = None
+    deprecated_removal_hint: ClassVar[Optional[str]] = None
 
+    # This method is marked @final so that we have a common deprecation mechanism, and to ensure
+    # that all subclasses have this exact type signature for their constructor.
     @final
     def __init__(self, raw_value: Optional[Any], *, address: Address) -> None:
+        self._check_deprecated(raw_value, address)
         # NB: We neither store the `address` or `raw_value` as attributes on this dataclass:
         # * Don't store `raw_value` because it very often is mutable and/or unhashable, which means
         #   this Field could not be passed around in the engine.
         # * Don't store `address` to avoid the cost in memory of storing `Address` on every single
         #   field encountered by Pants in a run.
```

Comment:
In retrospect, these comments are probably unnecessary. The docstring for `compute_value()` already explains the idea of needing to sanitize the raw value.

Maybe it's worth pointing out the `address` part, but I think it's fine without explanation. Where it's worth explaining is why we _do_ store it for `AsyncField`.